### DISCUSSION TITLE
Auth (and refresh) Difference between directus project and SDK with nextJS

### DISCUSSION THREAD

**Discussion by MrCasquette** [directus/directus#16677] [Archive]
Hi everybody. I have this problem for a long time ago and never managed to soved it totally. When i'm on directus dashboard My auth works perfectly with the refresh. I can see cookie etc. But with the SDK it works totally different,  I don't have a cookie except for the secure one I guess. but i have informations on the local storage. I can login and logout, but impossible to get the refresh token (good news for security) and it never works. And i see the refresh token is set with an hour of difference with my GMT.

Briefly I want to reproduce the exact directus dashboard behavior on my app.  I tried every config on same site, and cookie secure. Please tell me you have the answer 😅 It will help me so much

**Comment by paescuj** [comment]
Is there any specific reason why you need access to the refresh token? For security reasons the "refresh token" cookie is `HttpOnly` which means you won't be able to access it from client side.

If you really need access to it, you could set `auth.mode` to `json` (see https://docs.directus.io/reference/sdk.html#constructor) but I highly recommend to avoid this if possible.

**Reply by MrCasquette** [reply]
You are everywhere dude. 🤣 I think I misspoke. I just want to reproduce the exact behavior of my directus with my nextJS application which seems to have a problem with the refresh token. The login and logout is done without problems. But after 15 minutes of token ttl, the refresh does not work. Whereas in directus without going through the SDK so in native it works perfectly. I mean the refresh try but get an error.

**Reply by paescuj** [reply]
I see, thanks for the clarification!
Do you see any errors in the console when the login is happening and / or the token is refreshed?
Are you running your app on localhost? In that case you might run into the issue stated under https://docs.directus.io/reference/sdk.html#refresh-auth-token.
Where is your Directus instance running (localhost / server)?

**Reply by MrCasquette** [reply]
Every instance of directus I'm running are always online on a VPS. (I Tried two: mine and a friend one) I tried locally (I know this problem and solve it easily with config .env

> https://docs.directus.io/reference/sdk.html#refresh-auth-token.

But now i'm online on a vps (for example one of my project run on the same VPS but never disable the refresh) the other project run on an other vps (both directus and front -> nextJs) but the directus run on a sub sub domain. And this one often miss refresh

The errors occurs sometimes and other times (rarely) not. I don't know why. See:

**Reply by paescuj** [reply]
Hard to tell why it's not working sometimes...
As always it's advisable to keep an eye on the console!
Any chance that you've logged out in another tab? Note that you're "logged out" from your app too when you log out in the connected Directus instance (removes the refresh token cookie).

(You should better not post screenshots containing the refresh token as technically others could use it to login into your instance within the next 7 days: `REFRESH_TOKEN_TTL` - https://docs.directus.io/self-hosted/config-options.html#security)

**Reply by MrCasquette** [reply]
oh sh## I'm too tired these days. You're right!! Why i do that??? Thank for your time. All my hope went from your help, because it's a problem I have a long time ago. Do you personnaly work with react or next? Is this a problem that can come from that, unlike vueJs.

**Reply by paescuj** [reply]
Hey @CubW0lf, sorry for my late reply!

> Do you personnaly work with react or next? Is this a problem that can come from that, unlike vueJs.

Yes, I've worked with React and Next.js and it authentication incl. auto-refresh does work with the offical JS SDK, so this shouldn't be related.
Have you been able to solve the problem in the meantime?

**Reply by MrCasquette** [reply]
Not yet, unfortunately.But I'm interested in a basic and simple piece of code that I could possibly improve as I please since you have already worked with nextJs. It would be really cool to see what I did wrong. :)

**Reply by paescuj** [reply]
As already discussed on Discord, I can highly recommend the articles from Raphaël about using Directus together with Next.js, including this one which shows how to do authentication with the Directus SDK:
https://rphl.dev/blog/directus-sdk-nextjs-several-useswr (particularly the section under "Creating an AuthProvider using Directus js-sdk")
